File: src/patterns/behavioral/visitor/visitor.ts
// Visitor
interface IShapeVisitor {
  visitRectangle(rect: Rectangle): void;
  visitTriangle(tr: Triangle): void;
}

// Concrete Visitor
class AreaVisitor implements IShapeVisitor {
  visitRectangle(rect: Rectangle): void {
    console.log("The area of the rectangle is " + rect.width * rect.height);
  }

  visitTriangle(tr: Triangle): void {
    console.log("The area of the triangle is " + (tr.a * tr.b) / 2);
  }
}

// Components
interface IShape {
  getType(): string;
  accept(visitor: IShapeVisitor): void;
}

// Concrete Components
class Rectangle implements IShape {
  width: number;
  height: number;
  constructor(width: number, height: number) {
    this.width = width;
    this.height = height;
  }

  getType(): string {
    return "rectangle";
  }

  accept(visitor: IShapeVisitor): void {
    visitor.visitRectangle(this);
  }
}

class Triangle implements IShape {
  a: number;
  b: number;
  constructor(a: number, b: number) {
    this.a = a;
    this.b = b;
  }

  getType(): string {
    return "triangle";
  }

  accept(visitor: IShapeVisitor): void {
    visitor.visitTriangle(this);
  }
}

const rectangle = new Rectangle(10, 10);
const areaVisitor = new AreaVisitor();

rectangle.accept(areaVisitor);

File: src/patterns/behavioral/visitor/visitor.go
package main

import "fmt"

// Visitor
type IShapeVisitor interface {
	visitRectangle(r *Rectangle)
	visitTriangle(t *Triangle)
	visitCircle(c *Circle)
}

// Concrete Visitors
type AreaVisitor struct {

}

func (v *AreaVisitor) visitRectangle(r *Rectangle) {
	fmt.Printf("The area of this rectangle is %v", r.w * r.h)
}

func (v *AreaVisitor) visitTriangle(t *Triangle){
	fmt.Printf("The area of this triangle %v", (t.a * t.b ) / 2)
}

func (v *AreaVisitor) visitCircle(c *Circle){
	fmt.Printf("The area of this circle is %v", c.l * c.l)
}

// Component
type IShape interface {
	accept(IShapeVisitor)
	getType() string
}

// Concrete Components
type Rectangle struct {
	w int
	h int
}

func (r *Rectangle) getType() string {
	return "rectangle"
}

func (r *Rectangle) accept(visitor IShapeVisitor){
	visitor.visitRectangle(r)
}

type Circle struct {
	l int
}

func (c *Circle) getType() string {
	return "circle"
}

func (c *Circle) accept(visitor IShapeVisitor){
	visitor.visitCircle(c)
}

type Triangle struct {
	a int
	b int 
}

func (r *Triangle) getType() string {
	return "triangle"
}

func (t *Triangle) accept(visitor IShapeVisitor) {
	visitor.visitTriangle(t)
}

func main(){
	rect := &Rectangle{w: 10, h: 10}
	areaVisitor := &AreaVisitor{}

	rect.accept(areaVisitor)
}